### DISCUSSION TITLE
When is props destructuring safe and when we should use helpers

### DISCUSSION THREAD

**Discussion by jraoult** [solidjs/solid#408] [Q&A] [answered]
I recently stumbled upon Solid and it looks promising (my typical stack right now is TS+Mobx+React). I got started very quickly and now I'm at a point where I would expect some things to not work and they do. So clearly I don't fully understand the "re-rendering" phase.

I read these https://github.com/ryansolid/solid/blob/main/documentation/components.md#components and https://github.com/ryansolid/solid/blob/main/documentation/components.md#components and then I wrote this:

```ts
const OutlinedTextField = (
  props: {
    label?: JSX.Element;
    inputProps?: ComponentProps<"input">;
  } & ComponentProps<"label">
) => {
  const { label, inputProps, className, ...others } = props;

  // as the element executes the provided variable will be assigned
  let root: HTMLLabelElement | undefined;
  onMount(() => {
    if (root != null) {
      const mdcTextField = new MDCTextField(root);
      onCleanup(() => mdcTextField.destroy());
    }
  });

  return (
    <label
      className={cx("mdc-text-field", "mdc-text-field--outlined", className)}
      ref={root}
      {...others}
    >
      <span className="mdc-notched-outline">
        <span className="mdc-notched-outline__leading" />
        <span className="mdc-notched-outline__notch">
          {label && (
            <span className="mdc-floating-label" id="my-label-id">
              {label}
            </span>
          )}
        </span>
        <span className="mdc-notched-outline__trailing" />
      </span>
      <input
        type="text"
        className="mdc-text-field__input"
        aria-labelledby="my-label-id"
        {...inputProps}
      />
    </label>
  );
};
```

- I'm wondering if I should use the props helpers to destructure them? Would that change anything?
- `inputProps` contains the `value` prop that is controlled one level higher. What part exactly re-renders when `value` changes? Is it the "proper" (efficient) way to write this or is it a better way minimising updates?

**Comment by jraoult** [comment]
Hmm interesting. I actually realised that by destructuring props this way, the component doesn't react to changes made to `inputProps` upstream and by doing:

```ts
 const [local, others] = splitProps(props, [
    "label",
    "inputProps",
    "className",
  ]);
```
it actually works as intended.

So i guess my revised question is, **is there any case where it is safe/ok to use destructuring over `splitProps`**?

**Comment by ryansolid** [comment, marked as answer]
If it is inside a computation closure like in `createEffect` or `createMemo`.  Otherwise you will be doing so in an untracked context. These are the only things other than the JSX that re-run in Solid. You can view JSX as a number effects that update the DOM.

Since you are familiar with MobX this may be easier to follow. Solid's whole system is based on the reactive system. There is no other mechanism for rendering. So you can view everything as a bunch of nested `autorun`'s and `computed`'s to use the MobX terms. Basically, the DOM is a side effect. Since there is no VDOM we are generating real DOM nodes so to prevent unnecessary work I'm wrapping expressions in `computed`s.

The end result though is a system that the components are really nothing but function calls. They aren't wrapped or ever re-run. Since I don't want the parent context to track internals I `untrack` the body of every component. And from there everything works the way a reactive system works on its own.

This guide on MobX is helpful and applies here: https://doc.ebichu.cc/mobx/best/react.html

**Reply by jraoult** [reply]
thx @ryansolid for being on it 😀. I'm adjusting my mental model after 3 years of Mobx+React.

I think the thing I was missing is that the JSX transformation part is not just passing the property the downstream components but it does a little more. I assumed it was just generating something like that:
```js
untrack(() => Component({
  prop1: "static",
  prop2: state.dynamic 👈
}))
```
I couldn't figure why destructuring earlier would be any different and since I understood the body of the component's function was untracked by design I was very confused that dotted access would even be reactive at all. Hence my remark: _"I would expect some things to not work and they do"_.

But then I discovered as shown here https://github.com/ryansolid/solid/blob/main/documentation/components.md#props that it writes something like that instead:
```js
untrack(() => Component({
  prop1: "static",
  // dynamic expression so we wrap in a getter
  get prop2() { return state.dynamic }
}))
```
and it is therefore a chain of lately executed/evaluated accessors.

**Reply by ryansolid** [reply]
Exactly everything is lazily evaluated. The same rules that affect property tracking in MobX apply here. It is just in React they wrap the whole component in an autorun and we don't want to do that since it redoes all the work we've done. Things like hooks allow React to persist state over each run, but Solid's model is a lot simpler in that we already have a mechanism to re-run things so no need to re-run the component itself.

I enjoy this lazy evaluation model since you can start visualizing how we are more less flattening the component graph. Sure there are a few extra function calls (with the getters) but we aren't binding anything or creating new computations and in the end it is the final side effect (autorun so to speak) that does the heavy lifting. Still our props do run through our components so this allow for transformation while structurally we flatten along the control flows (if, for) the data itself can be viewed almost like a stream.

This is one of my favorite examples for this: https://codesandbox.io/s/granular-rendering-bs1pu?file=/index.js

**Reply by akotlar** [reply]
What are the disadvantage of passing the reactive object itself, rather than the result of calling the getter, meaning:

```
<Component id={id}/>
```
instead of
```
<Component id={id()}/>
```

Will Component re-render in the first case?

**Reply by snnsnn** [reply]
@akotlar There is none, compiler is smart enough to adapt accordingly. If it is an accessor, the compiled component will call the accessor, if it is an invocation, it will use getter so that reactivity is preserved.

```typescript
_$createComponent(Counter, {
  get value() {
    return count();
  }
})
```

```typescript
_$createComponent(Counter, {
  value: count
})
```

**Reply by akotlar** [reply]
Thanks! I have a seemingly confusing issue, given the expected behavior you outlined:

```
function UserHeader({user, onLogoutClicked}: userHeaderProps) {
  console.info("user in UserHeader", user, !!user)
  return (
    <Switch fallback={<A href="/login">Login</A>}>
      <Match when={user}>
        <A href="/user">
          <b>{user!.name}</b>
        </A>
        <button onclick={onLogoutClicked}>Log out</button>
      </Match>
    </Switch>
  );
};

function Header() {
  const [user, setUser] = createSignal(idToken.decodedToken ? idToken.decodedToken : null);

  const triggerLogout = () => idToken.logout();

  let loggedInCallback: CallbackFunction | undefined;

  let loggedOutCallback: CallbackFunction | undefined;

  loggedInCallback = addCallback(loggedInEventName, ([user_data, _]) => {
    console.info("got logged in event in header")
    console.info("user_data", user_data);
    setUser(user_data);
    console.info('user', user());
  });

  loggedOutCallback = addCallback(loggedOutEventName, () => {
    console.info("got logged out event in header")
    setUser(null);
    console.info('user', user());
  });

  onCleanup(() => {
    if (!loggedInCallback || !loggedOutCallback) {
      console.error("Header is missing expected authentication callbacks on cleanup. Check lifecycle methods");
      return;
    }

    removeCallback(loggedInEventName, loggedInCallback);
    removeCallback(loggedOutEventName, loggedOutCallback);
  });

  return (
      <span id="header">
        <A class="site-button--colored link md-button md-ink-ripple" href="/" title="/">Home</A>
        <A
          href="/submit"
          title="Submit"
        >Submit</A>
        <A
          href={`/${JobTypes.incomplete}`}
          title="Incomplete"
          end={true}
        >Incomplete</A>
        <A
          href={`/${JobTypes.failed}`}
          title="Failed"
        >Failed</A>
        <A
          href={`/${JobTypes.completed}`}
          title="Results"
        >Results</A>
        <A
          href={`/${JobTypes.public}`}
          title="Public"
        >Public</A>
        <A
          href={`/${JobTypes.shared}`}
          title="Shared"
        >Shared</A>

        <span id="profile-divider" />
        <UserHeader user={user()} onLogoutClicked={triggerLogout} />
      </span>
  );
};
```

In the above, clicking on onLogoutClicked triggers logout, which in turn triggers the callbacks attached to either loggedInEventName or loggedOutEventName. I can confirm this by examining console, which prints the expected debug messages. However, UserHeader is not updated, neither in the component body (the console.log is not triggered), nor in its returned JSX. I've verified that this has nothing to do with Switch.

However, if I pass the user getter function instead ( `{user}` ), UserHeader re-renders its JSX (the body of the function excluding the returned JSX is not re-executed, as expected). 

Could you help me understand this behavior?
